File: day1/day1.R
# Load Libraries ----------------------------------------------------------
library(R6)

# Load Data ---------------------------------------------------------------
in_data <- readLines("advent2025/day1/data/input.txt", warn=FALSE)

# Part 1 ------------------------------------------------------------

### Define Object
dialClass <- R6Class("dialClass",
 private = list(
   start_position = NULL,
   current_position = NULL,
   zero_count = NULL
 ),
 public = list(
   initialize = function(startPos) {
     private$start_position <- startPos
     private$current_position <- startPos
     private$zero_count <- 0
   },
   
   get_current_position = function() {
     private$current_position
   },
   
   get_zero_count = function() {
     private$zero_count
   },
   
   turn_dial = function(rotation) {
     direction <- substr(rotation,1,1)
     distance <- as.numeric(substr(rotation,2,nchar(rotation)))
     
     if (direction == "L") {
       new_position <- (private$current_position - distance) %% 100
     } else if (direction == "R") {
       new_position <- (private$current_position + distance) %% 100
     } else {
       print("Undefined Direction")
     }
     
     private$current_position <- new_position
   },
   
   process_turns = function(rotations) {
     for (rotation in rotations) {
       self$turn_dial(rotation)
       
       if (private$current_position == 0) {
         private$zero_count <- private$zero_count + 1
       }
     }
   }
 )
)

### Initialize Dial
dial <- dialClass$new(50)

dial$process_turns(in_data)

print(paste0("Part 1 Answer: ", dial$get_zero_count()))


# Part 2 ------------------------------------------------------------------
### Define Object
dialClass2 <- R6Class("dialClass2",
                     private = list(
                       start_position = NULL,
                       current_position = NULL,
                       zero_count = NULL
                     ),
                     public = list(
                       initialize = function(startPos) {
                         private$start_position <- startPos
                         private$current_position <- startPos
                         private$zero_count <- 0
                       },
                       
                       get_current_position = function() {
                         private$current_position
                       },
                       
                       get_zero_count = function() {
                         private$zero_count
                       },
                       
                       turn_dial = function(rotation) {
                         direction <- substr(rotation,1,1)
                         distance <- as.numeric(substr(rotation,2,nchar(rotation)))
                         
                         if (direction == "L") {
                           ###Convert to R (mod 100 in case position is 0)
                           position_norm <- (100 - private$current_position) %% 100
                           
                           new_position <- (private$current_position - distance) %% 100
                         } else if (direction == "R") {
                           position_norm <- private$current_position
                           
                           new_position <- (private$current_position + distance) %% 100
                         } else {
                           print("Undefined Direction")
                         }
                         
                         zeros <- floor((position_norm + distance) / 100)
                        
                         private$zero_count <-  private$zero_count + zeros
                         
                         private$current_position <- new_position
                       },
                       
                       process_turns = function(rotations) {
                         for (rotation in rotations) {
                           self$turn_dial(rotation)
                         }
                       }
                     )
)

### Initialize Dial
dial <- dialClass2$new(50)

dial$process_turns(in_data)

print(paste0("Part 2 Answer: ", dial$get_zero_count()))

File: day1/day1.rs
use std::fs;
use anyhow::Result;

#[derive(Debug)]
struct Rotation {
    direction: char,
    distance: isize,
}

struct Dial {
    start_position: isize,
    current_position: isize,
    zero_count_part1: isize,
    zero_count_part2: isize,
}

impl Dial {
    fn new(start_position: isize) -> Self {
        Dial {
            start_position: start_position,
            current_position: start_position,
            zero_count_part1: 0,
            zero_count_part2: 0,
        }
    }

    fn turn_dial(&mut self, rotation: &Rotation) {
        let mut position_norm: isize = self.current_position;

        if rotation.direction == 'L' {
            position_norm = (100 - self.current_position).rem_euclid(100);
            self.current_position = (self.current_position - rotation.distance).rem_euclid(100);
        } else if rotation.direction == 'R' {
            position_norm = self.current_position;
            self.current_position = (self.current_position + rotation.distance).rem_euclid(100);
        }

        if self.current_position == 0 {
            self.zero_count_part1 += 1;
        }

        let zeros: isize = (position_norm + rotation.distance) / 100;

        self.zero_count_part2 += zeros;
    }
}

fn main() -> Result<()> {
    let file_name = "data/input.txt";

    // Read in data
    let data_in = fs::read_to_string(file_name)?;

    // Parse Instructions into rotation struct
    let lines = data_in.split('\n');

    let mut rotations: Vec<Rotation> = Vec::new();

    for line in lines {
        let (direction, distance) = line.split_at(1);

        rotations.push(Rotation{
            direction: direction.parse::<char>().unwrap(),
            distance: distance.parse::<isize>().unwrap(),
        });
    }

    let mut my_dial: Dial = Dial::new(50);

    for rotation in rotations {
        my_dial.turn_dial(&rotation);
    }

    println!("Part 1 Answer: {}", my_dial.zero_count_part1);
    println!("Part 2 Answer: {}", my_dial.zero_count_part2);

    Ok(())
}
